crashtracker/src/crash_handler.rs

```diff
@@ -66,89 +62,133 @@ fn make_receiver(
             &config.path_to_receiver_binary
         ))?;
 
-    // Write the args into the receiver.
-    // Use the pipe to avoid secrets ending up on the commandline
-    writeln!(
-        receiver.stdin.as_ref().unwrap(),
-        "{}",
-        serde_json::to_string(&config)?
-    )?;
-    writeln!(
-        receiver.stdin.as_ref().unwrap(),
-        "{}",
-        serde_json::to_string(&metadata)?
-    )?;
+    update_metadata(metadata)?;
+    update_config(config)?;
+
     Ok(receiver)
 }
 
+/// Updates the crashtracker metadata for this process
+/// Metadata is stored in a global variable and sent to the crashtracking
+/// receiver when a crash occurs.
+///
+/// PRECONDITIONS:
+///     None
+/// SAFETY:
+///     Crash-tracking functions are not guaranteed to be reentrant.
+///     No other crash-handler functions should be called concurrently.
+/// ATOMICITY:
+///     This function uses a swap on an atomic pointer.
+pub fn update_metadata(metadata: CrashtrackerMetadata) -> anyhow::Result<()> {
+    let metadata_string = serde_json::to_string(&metadata)?;
+    let box_ptr = Box::into_raw(Box::new((metadata, metadata_string)));
+    let old = METADATA.swap(box_ptr, SeqCst);
+    if !old.is_null() {
+        // Safety: This can only come from a box above.
+        unsafe {
+            std::mem::drop(Box::from_raw(old));
+        }
+    }
+    Ok(())
+}
+
+/// Updates the crashtracker config for this process
+/// Config is stored in a global variable and sent to the crashtracking
+/// receiver when a crash occurs.
+///
+/// PRECONDITIONS:
+///     None
+/// SAFETY:
+///     Crash-tracking functions are not guaranteed to be reentrant.
+///     No other crash-handler functions should be called concurrently.
+/// ATOMICITY:
+///     This function uses a swap on an atomic pointer.
+pub fn update_config(config: CrashtrackerConfiguration) -> anyhow::Result<()> {
+    let config_string = serde_json::to_string(&config)?;
+    let box_ptr = Box::into_raw(Box::new((config, config_string)));
+    let old = CONFIG.swap(box_ptr, SeqCst);
+    if !old.is_null() {
+        // Safety: This can only come from a box above.
+        unsafe {
+            std::mem::drop(Box::from_raw(old));
+        }
+    }
+    Ok(())
+}
+
 pub fn setup_receiver(
-    config: &CrashtrackerConfiguration,
-    metadata: &CrashtrackerMetadata,
+    config: CrashtrackerConfiguration,
+    metadata: CrashtrackerMetadata,
 ) -> anyhow::Result<()> {
-    let new_receiver = make_receiver(config, metadata)?;
-    let old_receiver =
-        unsafe { std::mem::replace(&mut RECEIVER, GlobalVarState::Some(new_receiver)) };
+    let new_receiver = Box::into_raw(Box::new(make_receiver(config, metadata)?));
+    let old_receiver = RECEIVER.swap(new_receiver, SeqCst);
```

Comment:
Really minor: Since outside callers only see `setup_receiver` this note is more of a code arrangement thing.

These two lines

```
    let new_receiver = Box::into_raw(Box::new(make_receiver(config, metadata)?));
    let old_receiver = RECEIVER.swap(new_receiver, SeqCst);
```

suggest to me that the receiver is being set up atomically. Which in fact isn't the case -- the metadata and config are "global" even though they don't look like it.

So I'd suggest maybe not calling `update_metadata`/`update_config` directly in `make_receiver`, and instead doing something like:

```rust
update_config(config);
update_metadata(metadata);

let new_receiver = Box::into_raw(Box::new(make_receiver()?));
let old_receiver = RECEIVER.swap(new_receiver, SeqCst);
```

here directly?

Reply:
done